test/gtest/helpers.cpp

```diff
@@ -40,18 +53,36 @@ std::string makeP4Source(const char* file, unsigned line, const char* rawSource)
     return source.str();
 }
 
+std::string makeP4Source(const char* file, unsigned line, const char* rawSource) {
+    return makeP4Source(file, line, P4Headers::NONE, rawSource);
+}
+
 }  // namespace detail
 
-// A macro which should be used by unit tests to define P4 source code. It adds
-// additional information to the source code to aid in debugging; see
-// makeP4Source for more information.
-#define P4_SOURCE(SRC) detail::makeP4Source(__FILE__, __LINE__, SRC)
-
-/* preprocessing by prepending the content of core.p4 to test program */
-std::string with_core_p4(const std::string& pgm) {
-    std::ifstream input("p4include/core.p4");
-    std::stringstream sstr;
-    while (input >> sstr.rdbuf()) continue;
-    return sstr.str() + pgm;
+/* static */ P4CTestEnvironment* P4CTestEnvironment::get() {
+    static P4CTestEnvironment* instance = new P4CTestEnvironment;
+    return instance;
 }
 
+void P4CTestEnvironment::SetUp() {
+    auto readHeader = [](const char* filename) {
+        SCOPED_TRACE(filename);
+
+        std::ifstream input(filename);
+        EXPECT_TRUE(input.good());
+
+        // Initialize a buffer with a #line preprocessor directive. This ensures
+        // that any errors we encounter in this header will reference the
+        // correct file and line.
+        std::stringstream buffer;
+        buffer << "#line 1 \"" << filename << "\"" << std::endl;
+
+        // Read the header into the buffer and return it.
+        while (input >> buffer.rdbuf()) continue;
+        return buffer.str();
+    };
+
+    // XXX(seth): We should find a more robust way to locate these headers.
+    _coreP4 = readHeader("p4include/core.p4");
+    _v1Model = readHeader("p4include/v1model.p4");
```

Comment:
shouldn't these be predicated on what the test requested? You have an enum in the header file saying " Specifies which standard headers should be included by a GTest."